flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -103,6 +104,10 @@
 
     private final Time rpcTimeout;
 
+    private final Map<AllocationID, Long> reservedAllocationRefCounters;
```

Comment:
Did you introduce this map for performance reasons? Alternatively we could probably also get the same result via `new HashSet<>(reservedAllocationsByExecutionVertex.values())`.

Reply:
Using `new HashSet<>(reservedAllocationsByExecutionVertex.values())` will introduce O(V) complexity. And invoking it V times means a O(V^2) complexity. (V is the number of tasks.)
`reservedAllocationRefCounters.keySet()` is O(1) because it just creates a wrapper.

Reply:
Sounds good. Thanks for the explanation.